File: cirq/basic/logic-gates/logic_gates.py
"""Single-qubit logic gates: X, Y, Z, H, S, T with state transformations."""

import numpy as np
import cirq


def main() -> None:
    q = cirq.LineQubit(0)
    sim = cirq.Simulator()

    gates: list[tuple[str, cirq.Gate, list[cirq.Operation]]] = [
        ("X (bit-flip)", cirq.X, [cirq.X(q)]),
        ("Y", cirq.Y, [cirq.Y(q)]),
        ("Z (phase-flip)", cirq.Z, [cirq.Z(q)]),
        ("H (Hadamard)", cirq.H, [cirq.H(q)]),
        ("S (√Z)", cirq.S, [cirq.S(q)]),
        ("T (√S)", cirq.T, [cirq.T(q)]),
    ]

    print("=== Starting from |0⟩ ===\n")
    for name, gate, ops in gates:
        circuit = cirq.Circuit(ops)
        result = sim.simulate(circuit)
        sv = result.final_state_vector
        print(f"{name}:")
        print(f"  Matrix:\n{np.array2string(np.array(gate.unitary()), precision=3)}")
        print(f"  State: {np.array2string(sv, precision=4)}")
        print(f"  |0⟩ amp: {sv[0]:.4f},  |1⟩ amp: {sv[1]:.4f}")
        print()

    print("=== Combining gates: T† S H |0⟩ ===")
    circuit = cirq.Circuit([cirq.H(q), cirq.S(q), cirq.T(q) ** -1])
    result = sim.simulate(circuit)
    print(circuit)
    print(f"State: {result.final_state_vector}")


if __name__ == "__main__":
    main()

File: qclojure/basic/logic_gates/logic_gates.clj
(ns basic.logic-gates
  (:require [org.soulspace.qclojure.domain.circuit :as qc]
            [org.soulspace.qclojure.application.backend :as qb]
            [org.soulspace.qclojure.adapter.backend.ideal-simulator :as sim]))

(defn run-gate [backend gate-name gate-fn]
  (let [circuit (-> (qc/create-circuit 1 gate-name "Single-qubit gate demo")
                    (gate-fn 0))
        result  (qb/execute-circuit backend circuit {:shots 1000})
        probs   (:probabilities (:final-state result))]
    (println (str "  " gate-name " on |0⟩:"))
    (doseq [[basis prob] (sort-by key probs)]
      (printf "    |%s⟩  %.4f%n" basis prob))
    (println)))

(defn -main []
  (println "=== Logic Gates ===")
  (println)
  (let [backend (sim/create-simulator)]
    (run-gate backend "X (Pauli-X)" qc/x-gate)
    (run-gate backend "Y (Pauli-Y)" qc/y-gate)
    (run-gate backend "Z (Pauli-Z)" qc/z-gate)
    (run-gate backend "H (Hadamard)" qc/h-gate)
    (run-gate backend "S (Phase)"   qc/s-gate)
    (run-gate backend "T (π/8)"    qc/t-gate)

    (println "Gate effects on |0⟩:")
    (println "  X: |0⟩ → |1⟩  (bit flip)")
    (println "  Y: |0⟩ → i|1⟩")
    (println "  Z: |0⟩ → |0⟩  (phase flip, no effect on |0⟩)")
    (println "  H: |0⟩ → |+⟩  (superposition)")
    (println "  S: |0⟩ → |0⟩  (phase gate, no effect on |0⟩)")
    (println "  T: |0⟩ → |0⟩  (π/8 gate, no effect on |0⟩)")))
